In [365]:
import numpy as np
import pandas as pd
from statistics import mean

In [366]:
life_df = pd.read_csv('Life Expectancy Data.csv', delimiter=',')

In [367]:
# Remove whitespaces before & after column name and capitalise all column names.
life_df.columns = life_df.columns.str.strip()
life_df.columns = life_df.columns.str.title()
life_df = life_df.convert_dtypes()

In [368]:
life_df.dtypes

Country                             string
Year                                 Int64
Status                              string
Life Expectancy                    Float64
Adult Mortality                      Int64
Infant Deaths                        Int64
Alcohol                            Float64
Percentage Expenditure             Float64
Hepatitis B                          Int64
Measles                              Int64
Bmi                                Float64
Under-Five Deaths                    Int64
Polio                                Int64
Total Expenditure                  Float64
Diphtheria                           Int64
Hiv/Aids                           Float64
Gdp                                Float64
Population                         Float64
Thinness  1-19 Years               Float64
Thinness 5-9 Years                 Float64
Income Composition Of Resources    Float64
Schooling                          Float64
dtype: object

In [369]:
life_df

Country  Year      Status  Life Expectancy  Adult Mortality  \
0     Afghanistan  2015  Developing             65.0              263   
1     Afghanistan  2014  Developing             59.9              271   
2     Afghanistan  2013  Developing             59.9              268   
3     Afghanistan  2012  Developing             59.5              272   
4     Afghanistan  2011  Developing             59.2              275   
...           ...   ...         ...              ...              ...   
2933     Zimbabwe  2004  Developing             44.3              723   
2934     Zimbabwe  2003  Developing             44.5              715   
2935     Zimbabwe  2002  Developing             44.8               73   
2936     Zimbabwe  2001  Developing             45.3              686   
2937     Zimbabwe  2000  Developing             46.0              665   

      Infant Deaths  Alcohol  Percentage Expenditure  Hepatitis B  Measles  \
0                62     0.01               71.279624           65     1154   
1                64     0.01               73.523582           62      492   
2                66     0.01               73.219243           64      430   
3                69     0.01               78.184215           67     2787   
4                71     0.01                7.097109           68     3013   
...             ...      ...                     ...          ...      ...   
2933             27     4.36                     0.0           68       31   
2934             26     4.06                     0.0            7      998   
2935             25     4.43                     0.0           73      304   
2936             25     1.72                     0.0           76      529   
2937             24     1.68                     0.0           79     1483   

      ...  Polio  Total Expenditure  Diphtheria  Hiv/Aids         Gdp  \
0     ...      6               8.16          65       0.1   584.25921   
1     ...     58               8.18          62       0.1  612.696514   
2     ...     62               8.13          64       0.1  631.744976   
3     ...     67               8.52          67       0.1     669.959   
4     ...     68               7.87          68       0.1   63.537231   
...   ...    ...                ...         ...       ...         ...   
2933  ...     67               7.13          65      33.6  454.366654   
2934  ...      7               6.52          68      36.7  453.351155   
2935  ...     73               6.53          71      39.8    57.34834   
2936  ...     76               6.16          75      42.1  548.587312   
2937  ...     78                7.1          78      43.5  547.358878   

      Population  Thinness  1-19 Years  Thinness 5-9 Years  \
0     33736494.0                  17.2                17.3   
1       327582.0                  17.5                17.5   
2     31731688.0                  17.7                17.7   
3      3696958.0                  17.9                18.0   
4      2978599.0                  18.2                18.2   
...          ...                   ...                 ...   
2933  12777511.0                   9.4                 9.4   
2934  12633897.0                   9.8                 9.9   
2935    125525.0                   1.2                 1.3   
2936  12366165.0                   1.6                 1.7   
2937  12222251.0                  11.0                11.2   

      Income Composition Of Resources  Schooling  
0                               0.479       10.1  
1                               0.476       10.0  
2                                0.47        9.9  
3                               0.463        9.8  
4                               0.454        9.5  
...                               ...        ...  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427       10.0  
2936                            0.427        9.8  
2937         

In [370]:
# Total null values amongst each column.
print(life_df.isnull().sum())

Country                              0
Year                                 0
Status                               0
Life Expectancy                     10
Adult Mortality                     10
Infant Deaths                        0
Alcohol                            194
Percentage Expenditure               0
Hepatitis B                        553
Measles                              0
Bmi                                 34
Under-Five Deaths                    0
Polio                               19
Total Expenditure                  226
Diphtheria                          19
Hiv/Aids                             0
Gdp                                448
Population                         652
Thinness  1-19 Years                34
Thinness 5-9 Years                  34
Income Composition Of Resources    167
Schooling                          163
dtype: int64


# Total Expenditure Cleaning
By Anthony George 220180567
##### Aim: Complete the Total Expenditure column dataset without applying bias
The cleaning of *Total Expenditure* column data is two-fold. <br>
 1. Zero's missing, causing values of '10' to be values of '1', biasing the data terribly. <br>
 2. Values for *Year = 2015* are missing throughout this dataset column<br>

To correct the 'zero's missing' value error, I will develop an algorithm to add a zero where needed.<br>
To correct the *Year* value error, since the percentage of missing data is low,  we can use the **imputation method** to complete the dataset of this column. 
We do this by taking an average of all *Total Expenditure* values for each *Country* over all other *Years*, and populate the empty field with that value. This way, the missing data can be predicted based on the complete observed data.
The imputation method implies the development of reasonable guesses for missing data.  If the portion of missing data is too high, which in this case it is not, the results lack natural variation that could result in an effective model.

Reference: https://www.mastersindatascience.org/learning/how-to-deal-with-missing-data/

**First, let's check which countries have a *Total Expenditure* of greater than zero for the year 2015**

In [371]:
print("\nThe following countries have a Total Expenditure > 0:\n")
print(life_df['Country'].loc[(life_df['Year'] == 2015) & (life_df['Total Expenditure'] > 0)].to_string(index=False))


The following countries have a Total Expenditure > 0:

Afghanistan
    Albania


#### Only two countries!
#### This means that all countries (excluding Afghanistan and Albania) have a Total Expenditure null value for Year = 2015 
Assessing the missing values, we can check the number of unique countries against the number of countries that have a 'Year' column value of '2015'

In [372]:
print("All Unique Countries: ", life_df['Country'].nunique())
print("Countries when Year=2015:", len(life_df.loc[life_df['Year'] == 2015]))

All Unique Countries:  193
Countries when Year=2015: 183


From these results, we can see there are 10 countries that do NOT contain data for the year 2015.
We should check the rest of the data of these countries to assess their validity 

In [373]:
# the first 'Year' entry for each country should be '2015', so let's check that.
country = ''
country_list = []
for index, row in life_df.iterrows():
    if country != row['Country']:
        country = row['Country']
        if row['Year'] == 2015:
            continue
        else:
            country_list.append(row)

print("\nThe following countries do not have data for any 'Year' other than '2013'\n")
df = pd.DataFrame(country_list)
df


The following countries do not have data for any 'Year' other than '2013'



Country  Year      Status Life Expectancy Adult Mortality  \
624            Cook Islands  2013  Developing            <NA>            <NA>   
769                Dominica  2013  Developing            <NA>            <NA>   
1650       Marshall Islands  2013  Developing            <NA>            <NA>   
1715                 Monaco  2013  Developing            <NA>            <NA>   
1812                  Nauru  2013  Developing            <NA>            <NA>   
1909                   Niue  2013  Developing            <NA>            <NA>   
1958                  Palau  2013  Developing            <NA>            <NA>   
2167  Saint Kitts and Nevis  2013  Developing            <NA>            <NA>   
2216             San Marino  2013  Developing            <NA>            <NA>   
2713                 Tuvalu  2013  Developing            <NA>            <NA>   

      Infant Deaths Alcohol  Percentage Expenditure  Hepatitis B  Measles  \
624               0    0.01                0.000000           98        0   
769               0    0.01               11.419555           96        0   
1650              0    0.01              871.878317            8        0   
1715              0    0.01                0.000000           99        0   
1812              0    0.01               15.606596           87        0   
1909              0    0.01                0.000000           99        0   
1958              0    <NA>              344.690631           99        0   
2167              0    8.54                0.000000           97        0   
2216              0    0.01                0.000000           69        0   
2713              0    0.01               78.281203            9        0   

      ... Polio  Total Expenditure  Diphtheria  Hiv/Aids          Gdp  \
624   ...    98               3.58          98       0.1         <NA>   
769   ...    96               5.58          96       0.1    722.75665   
1650  ...    79              17.24          79       0.1  3617.752354   
1715  ...    99               4.30          99       0.1         <NA>   
1812  ...    87               4.65          87       0.1    136.18321   
1909  ...    99               7.20          99       0.1         <NA>   
1958  ...    99               9.27          99       0.1   1932.12237   
2167  ...    96               6.14          96       0.1         <NA>   
2216  ...    69               6.50          69       0.1         <NA>   
2713  ...     9              16.61           9       0.1   3542.13589   

      Population Thinness  1-19 Years Thinness 5-9 Years  \
624         <NA>                  0.1                0.1   
769         <NA>                  2.7                2.6   
1650        <NA>                  0.1                0.1   
1715        <NA>                 <NA>               <NA>   
1812        <NA>                  0.1                0.1   
1909        <NA>                  0.1                0.1   
1958       292.0                  0.1                0.1   
2167        <NA>                  3.7                3.6   
2216        <NA>                 <NA>               <NA>   
2713      1819.0                  0.2                0.1   

     Income Composition Of Resources Schooling  
624                             <NA>      <NA>  
769                            0.721      12.7  
1650                            <NA>       0.0  
1715                            <NA>      <NA>  
1812                            <NA>       9.6  
1909                            <NA>      <NA>  
1958                           0.779      14.2  
2167                           0.749      13.4  
2216                            <NA>      15.1  
2713                            <NA>       0.0  

[10 rows x 22 columns]

As we can see above, these 10 countries have some data populated for the *Year* 2013 only, with the zero data *Life Expectancy*. Since this column is crucial to the remainder of our analysis, **the removal of these entries from the dataset merits justification.**

In [374]:
# Drop the above rows from the dataset
for index, row in df.iterrows():
    life_df.drop(index, inplace=True)

life_df

Country  Year      Status  Life Expectancy  Adult Mortality  \
0     Afghanistan  2015  Developing             65.0              263   
1     Afghanistan  2014  Developing             59.9              271   
2     Afghanistan  2013  Developing             59.9              268   
3     Afghanistan  2012  Developing             59.5              272   
4     Afghanistan  2011  Developing             59.2              275   
...           ...   ...         ...              ...              ...   
2933     Zimbabwe  2004  Developing             44.3              723   
2934     Zimbabwe  2003  Developing             44.5              715   
2935     Zimbabwe  2002  Developing             44.8               73   
2936     Zimbabwe  2001  Developing             45.3              686   
2937     Zimbabwe  2000  Developing             46.0              665   

      Infant Deaths  Alcohol  Percentage Expenditure  Hepatitis B  Measles  \
0                62     0.01               71.279624           65     1154   
1                64     0.01               73.523582           62      492   
2                66     0.01               73.219243           64      430   
3                69     0.01               78.184215           67     2787   
4                71     0.01                7.097109           68     3013   
...             ...      ...                     ...          ...      ...   
2933             27     4.36                     0.0           68       31   
2934             26     4.06                     0.0            7      998   
2935             25     4.43                     0.0           73      304   
2936             25     1.72                     0.0           76      529   
2937             24     1.68                     0.0           79     1483   

      ...  Polio  Total Expenditure  Diphtheria  Hiv/Aids         Gdp  \
0     ...      6               8.16          65       0.1   584.25921   
1     ...     58               8.18          62       0.1  612.696514   
2     ...     62               8.13          64       0.1  631.744976   
3     ...     67               8.52          67       0.1     669.959   
4     ...     68               7.87          68       0.1   63.537231   
...   ...    ...                ...         ...       ...         ...   
2933  ...     67               7.13          65      33.6  454.366654   
2934  ...      7               6.52          68      36.7  453.351155   
2935  ...     73               6.53          71      39.8    57.34834   
2936  ...     76               6.16          75      42.1  548.587312   
2937  ...     78                7.1          78      43.5  547.358878   

      Population  Thinness  1-19 Years  Thinness 5-9 Years  \
0     33736494.0                  17.2                17.3   
1       327582.0                  17.5                17.5   
2     31731688.0                  17.7                17.7   
3      3696958.0                  17.9                18.0   
4      2978599.0                  18.2                18.2   
...          ...                   ...                 ...   
2933  12777511.0                   9.4                 9.4   
2934  12633897.0                   9.8                 9.9   
2935    125525.0                   1.2                 1.3   
2936  12366165.0                   1.6                 1.7   
2937  12222251.0                  11.0                11.2   

      Income Composition Of Resources  Schooling  
0                               0.479       10.1  
1                               0.476       10.0  
2                                0.47        9.9  
3                               0.463        9.8  
4                               0.454        9.5  
...                               ...        ...  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427       10.0  
2936                            0.427        9.8  
2937         

**Now we have a slightly cleaner dataset, we can check the exact number of rows we need to update with our coming algorithm**

In [375]:
print('Number of rows that require updating:', 
      len(life_df.loc[(life_df['Year'] == 2015) & (life_df['Total Expenditure'].isnull())]),'\n') 

print(life_df[['Country', 'Year', 'Total Expenditure']].loc
      [
    (life_df['Year'] == 2015) & 
    (life_df['Total Expenditure'].isnull())
      ])

Number of rows that require updating: 181 

                                 Country  Year  Total Expenditure
32                               Algeria  2015               <NA>
48                                Angola  2015               <NA>
64                   Antigua and Barbuda  2015               <NA>
80                             Argentina  2015               <NA>
96                               Armenia  2015               <NA>
...                                  ...   ...                ...
2858  Venezuela (Bolivarian Republic of)  2015               <NA>
2874                            Viet Nam  2015               <NA>
2890                               Yemen  2015               <NA>
2906                              Zambia  2015               <NA>
2922                            Zimbabwe  2015               <NA>

[181 rows x 3 columns]


**Algorithm to find and correct 'missing zeros' values that should be ten's, but are one's**

In [376]:
print('\nCompensated values:\n')
ctry = ''
for index, row in life_df.iterrows():
    if country != '' and pd.notnull(row['Total Expenditure']):
        if ctry != row['Country']:
            ctry = row['Country']
            flag = False
        if row['Total Expenditure'] > 9 and flag == False:
            for i in life_df.index:
                if life_df['Country'][i] == ctry:
                    flag = True
                    if pd.notnull(life_df['Total Expenditure'][i]):
                        row = str(life_df['Total Expenditure'][i])
                        num, dec = row.split('.')
                        if int(num) == 1:
                            num = num[:1] + '0' + num[1:]
                            num = num + '.' + dec
                            num = float(num)
                            
                            # update Total Expenditure column with updated values
                            life_df['Total Expenditure'][i] = num
                            print(i, ctry, life_df['Total Expenditure'][i])



Compensated values:

132 Austria 10.94
135 Austria 10.6
136 Austria 10.4
137 Austria 10.35
138 Austria 10.53
139 Austria 10.56
140 Austria 10.48
141 Austria 10.27
142 Austria 10.12
143 Austria 10.6
241 Belgium 10.59
242 Belgium 10.57
243 Belgium 10.54
244 Belgium 10.42
245 Belgium 10.17
246 Belgium 10.39
424 Burundi 10.3
497 Canada 10.45
498 Canada 10.67
499 Canada 10.78
500 Canada 10.82
503 Canada 10.3
661 Cuba 10.6
662 Cuba 10.19
664 Cuba 10.79
665 Cuba 10.42
738 Denmark 10.8
740 Denmark 10.98
741 Denmark 10.87
744 Denmark 10.18
754 Djibouti 10.57
937 France 10.57
938 France 10.45
939 France 10.53
940 France 10.6
941 France 10.53
942 France 10.43
943 France 10.22
983 Georgia 10.5
984 Georgia 10.19
997 Germany 10.99
998 Germany 10.93
1001 Germany 10.39
1002 Germany 10.18
1003 Germany 10.34
1004 Germany 10.52
1005 Germany 10.37
1006 Germany 10.62
1007 Germany 10.4
1008 Germany 10.15
1009 Germany 10.1
1126 Haiti 10.41
1182 Iceland 10.7
1315 Japan 10.23
1316 Japan 10.25
1317 Japan 10.17

**Now that we have the correct values in place, aggregation can be done to calculate a value to populate the *Total Expenditure* field of each *Country* for *Year* == 2015**

In [377]:
print('Average of all ')
country = ''
total = []
percentages = {}
for index, row in life_df.iterrows():
    if country != row['Country']:
        if country != '' and len(total)>0:
            average = mean(total)
            percentages[country] = round(average, 2)
        country = row['Country']
        total = []
    if row['Year'] == 2015:
        continue
    else:
        if pd.notnull(row['Total Expenditure']):
            total.append(row['Total Expenditure'])

for i in percentages:
    print(i, percentages[i])

Average of all 
Afghanistan 8.26
Albania 5.94
Algeria 4.6
Angola 3.92
Antigua and Barbuda 4.79
Argentina 6.91
Armenia 4.86
Australia 8.84
Austria 10.72
Azerbaijan 5.68
Bahamas 6.71
Bahrain 3.81
Bangladesh 2.85
Barbados 6.12
Belarus 6.19
Belgium 9.56
Belize 5.04
Benin 4.64
Bhutan 5.35
Bolivia (Plurinational State of) 5.63
Bosnia and Herzegovina 8.74
Botswana 5.51
Brazil 8.02
Brunei Darussalam 2.82
Bulgaria 7.24
Burkina Faso 6.04
Burundi 7.69
Côte d'Ivoire 5.72
Cabo Verde 4.76
Cambodia 5.74
Cameroon 4.72
Canada 10.05
Central African Republic 4.04
Chad 4.38
Chile 6.91
China 4.92
Colombia 6.34
Comoros 4.83
Congo 2.99
Costa Rica 8.62
Croatia 7.37
Cuba 8.83
Cyprus 6.76
Czechia 7.05
Democratic Republic of the Congo 3.56
Denmark 10.22
Djibouti 7.67
Dominican Republic 4.68
Ecuador 5.88
Egypt 5.35
El Salvador 7.1
Equatorial Guinea 2.93
Eritrea 3.47
Estonia 5.69
Ethiopia 4.9
Fiji 3.9
Finland 8.61
France 10.76
Gabon 3.18
Gambia 5.31
Georgia 8.53
Germany 10.67
Ghana 4.33
Greece 9.04
Grenada 6.19
Gu

Comparing the result of the aggregation with known results, we see that this imputation to derive the missing values for this column is actually quite accurate.<br><br>
We can see the aggregation has calculated Afghanistan has a Total Percentage for Year 2015 of 8.26 <br>
Compared with Afghanistan's actual Total Percentage for Year 2015 is 8.16 <br>
Similarly, Albania's actual = 6.0 vs the calcualted 5.94 <br>
Pretty accurate!<br>
Let's update the dataset

In [384]:
for i in percentages:     
    life_df.loc[(life_df['Country'] == i) & (life_df['Year'] == 2015), 'Total Expenditure'] = percentages[i]
    #print(life_df.loc[(life_df['Country'] == i) & (life_df['Year'] == 2015)])
pd.set_option('display.max_rows', 3000)
life_df.loc[life_df['Year'] == 2015]

Country  Year      Status  \
0                                           Afghanistan  2015  Developing   
16                                              Albania  2015  Developing   
32                                              Algeria  2015  Developing   
48                                               Angola  2015  Developing   
64                                  Antigua and Barbuda  2015  Developing   
80                                            Argentina  2015  Developing   
96                                              Armenia  2015  Developing   
112                                           Australia  2015   Developed   
128                                             Austria  2015   Developed   
144                                          Azerbaijan  2015  Developing   
160                                             Bahamas  2015  Developing   
176                                             Bahrain  2015  Developing   
192                                          Bangladesh  2015  Developing   
208                                            Barbados  2015  Developing   
224                                             Belarus  2015  Developing   
240                                             Belgium  2015   Developed   
256                                              Belize  2015  Developing   
272                                               Benin  2015  Developing   
288                                              Bhutan  2015  Developing   
304                    Bolivia (Plurinational State of)  2015  Developing   
320                              Bosnia and Herzegovina  2015  Developing   
336                                            Botswana  2015  Developing   
352                                              Brazil  2015  Developing   
368                                   Brunei Darussalam  2015  Developing   
384                                            Bulgaria  2015   Developed   
400                                        Burkina Faso  2015  Developing   
416                                             Burundi  2015  Developing   
432                                       Côte d'Ivoire  2015  Developing   
448                                          Cabo Verde  2015  Developing   
464                                            Cambodia  2015  Developing   
480                                            Cameroon  2015  Developing   
496                                              Canada  2015  Developing   
512                            Central African Republic  2015  Developing   
528                                                Chad  2015  Developing   
544                                               Chile  2015  Developing   
560                                               China  2015  Developing   
576                                            Colombia  2015  Developing   
592                                             Comoros  2015  Developing   
608                                               Congo  2015  Developing   
625                                          Costa Rica  2015  Developing   
641                                             Croatia  2015   Developed   
657                                                Cuba  2015  Developing   
673                                              Cyprus  2015   Developed   
689                                             Czechia  2015   Developed   
705               Democratic People's Republic of Korea  2015  Developing   
721                    Democratic Republic of the Congo  2015  Developing   
737                                             Denmark  2015   Developed   
753                                            Djibouti  2015  Developing   
770                                  Dominican Republic  2015  Developing   
786                                             Ecuador  2015  Developing   
802                                               Egypt  2015  Developing   
818                                         

In [30]:
# If all null are dropped.
#clean_df = life_df.dropna()

In [9]:
# 2015 percentage exp. 
# unique country names.